services/121-service/src/payments/fsp-integration/safaricom/safaricom.service.ts

```diff
@@ -0,0 +1,204 @@
+import { Injectable } from '@nestjs/common';
+import crypto from 'crypto';
+// import { SafaricomPaymentPayloadDto } from './dto/safaricom-payment-payload.dto';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { FspName } from '../../../fsp/enum/fsp-name.enum';
+import {
+  FspTransactionResultDto,
+  PaTransactionResultDto,
+} from '../../../payments/dto/payment-transaction-result.dto';
+import { TransactionEntity } from '../../../payments/transactions/transaction.entity';
+import { StatusEnum } from '../../../shared/enum/status.enum';
+import { PaPaymentDataDto } from '../../dto/pa-payment-data.dto';
+import { TransactionsService } from '../../transactions/transactions.service';
+import { SafaricomTransferPayload } from './dto/safaricom-transfer-payload.dto';
+import { SafaricomRequestEntity } from './safaricom-request.entity';
+import { SafaricomApiService } from './safaricom.api.service';
+
+@Injectable()
+export class SafaricomService {
+  @InjectRepository(SafaricomRequestEntity)
+  private readonly safaricomRequestRepository: Repository<SafaricomRequestEntity>;
+  @InjectRepository(TransactionEntity)
+  private readonly transactionRepository: Repository<TransactionEntity>;
+
+  public constructor(
+    private readonly safaricomApiService: SafaricomApiService,
+    private readonly transactionsService: TransactionsService,
+  ) {}
+
+  public async sendPayment(
+    paymentList: PaPaymentDataDto[],
+    programId: number,
+    paymentNr: number,
+  ): Promise<FspTransactionResultDto> {
+    const fspTransactionResult = new FspTransactionResultDto();
+    fspTransactionResult.paList = [];
+    fspTransactionResult.fspName = FspName.safaricom;
+
+    await this.safaricomApiService.authenticate();
```

Comment:
This may be good to move inside the for loop. Because if this for loop takes long to process, it could become an expired token

Reply:
You are right and I will move inside the loop.